do not skip Generator for MissingTypehintCheck

Flyingmana opened the pull request:
references [#7498](https://github.com/phpstan/phpstan/issues/7498)

only a draft for now, I want to see which test are all failing and for what reasons.

Flyingmana committed:
do not skip Generator for MissingTypehintCheck

ondrejmirtes commented:
Explained the wanted changes here https://github.com/phpstan/phpstan/issues/7498#issuecomment-3239381890

Flyingmana committed:
align MissingTypehintCheck constructs with new argument

VincentLanglet commented:
I would love to this moving forward. Are you still interested in your PR @Flyingmana ?

I think you just need to
- Rebase/Fix conflicts
- Describe the iterable type in some stubs/tests

Flyingmana commented:
Yes, might have time for it later today, but latest next weekend

Flyingmana committed:
do not skip generic iterables for MissingTypehintCheck on bleeding edge

# Conflicts:
#	conf/bleedingEdge.neon
#	conf/config.neon
#	conf/parametersSchema.neon

Flyingmana committed:
more specific return type in Stub PDOStatement::getIterator

Flyingmana committed:
align unit tests for new generic Iterable checks

Flyingmana commented on tests/PHPStan/Analyser/nsrt/bug-8886.php:
```diff
@@ -9,5 +9,5 @@ function testPDOStatementGetIterator(): void {
     $stmt = $pdo->query('SELECT 1');
 
-	assertType('Iterator', $stmt->getIterator());
+	assertType('Iterator<mixed, array<int|string, mixed>>', $stmt->getIterator());
```
not sure if there is an alternative method for the type check, but having a more explicit check is probably ok (I hope)

Flyingmana commented:
I think now all is fixed which belongs to this repository, the rest is separate repositories

ondrejmirtes commented on src/Rules/MissingTypehintCheck.php:
```diff
@@ -118,7 +120,10 @@ public function getNonGenericObjectTypesWithGenericClass(Type $type): array
 				}
-				if (in_array($classReflection->getName(), self::ITERABLE_GENERIC_CLASS_NAMES, true)) {
+				if (
+					$this->checkGenericIterableClasses !== true &&
```
Please do `!$this->checkGenericIterableClasses` instead

ondrejmirtes commented on tests/PHPStan/Rules/Functions/MissingFunctionParameterTypehintRuleTest.php:
```diff
@@ -71,6 +71,11 @@ public function testRule(): void
 				MissingTypehintCheck::MISSING_ITERABLE_VALUE_TYPE_TIP,
 			],
+			[
+				'Function MissingFunctionParameterTypehint\missingTraversableTypehint() has parameter $traversable with generic interface Traversable but does not specify its types: TKey, TValu',
```
Pretty sure this will fail with a typo - TValu vs. TValue

ondrejmirtes reviewed (changes requested):
I'd like to see more tests with the other no-longer-skipped types like Generator.

Flyingmana committed:
simplify boolean check

Flyingmana committed:
fix typo in expected error message

Flyingmana committed:
add test stubs for Generator Return Types

Flyingmana committed:
add expected reports of generator without specified types

Flyingmana committed:
cleanup duplicated entry and not exsting typehint tip

Flyingmana committed:
add more testcase for generic types

Flyingmana commented:
> I'd like to see more tests with the other no-longer-skipped types like Generator.

I added new tests/examples for generics in 
tests/PHPStan/Rules/Functions/data/missing-function-return-typehint.php

Flyingmana committed:
correct expected reported errors

Flyingmana committed:
change union type to intersectional type

ondrejmirtes commented on stubs/runtime/ReflectionAttribute.php:
```diff
@@ -22,10 +22,16 @@ public function isRepeated(): bool
 		}
 
+		/**
+		 * @return array<array-key, mixed>;
```
Why the semicolon?

Flyingmana replied:
fixed/removed

ondrejmirtes commented on stubs/runtime/ReflectionAttribute.php:
```diff
@@ -22,10 +22,16 @@ public function isRepeated(): bool
 		}
 
+		/**
+		 * @return self
```
What? That does not make sense.

Flyingmana replied:
no idea anymore, but you are correct, that does not make sense, have removed it again

ondrejmirtes reviewed (changes requested):
I may have forgotten some context here but why is this necessary when for example `Traversable` return type already gets `return type has no value type specified in iterable type Traversable`? https://phpstan.org/r/53c35e3b-1a18-40c9-9ca1-6199bdfa1ed5

Flyingmana commented:
> I may have forgotten some context here but why is this necessary when for example `Traversable` return type already gets `return type has no value type specified in iterable type Traversable`? https://phpstan.org/r/53c35e3b-1a18-40c9-9ca1-6199bdfa1ed5

might have slipped my mind that it was already covered properly when I worked on it the last time. Will remove this part again. (and fix the conflict, check the tests again)

Flyingmana committed:
remove typo and wrong self return type

Flyingmana committed:
skip traversable to avoid double detection

Flyingmana committed:
restore previous traversable related tests

Flyingmana committed:
extend detected error in test

Flyingmana commented:
another try, Hope its now good :) 
it seems the failing tests are all unrelated.
If more is to do, I currently have more time available in the evenings

ondrejmirtes reviewed (changes requested):
Thank you, this makes more sense to me now. Can you please add tests for a pre-existing `getIterableTypesWithMissingValueTypehint` function (for example in MissingMethodReturnTypehintRule) that would check all ITERABLE_GENERIC_CLASS_NAMES to make sure that each of them is either covered by getIterableTypesWithMissingValueTypehint or by the code you've written here?

I'm a little bit worried that the current differentiation for Traversable vs. the others is brittle so I want to be sure it doesn't break in the future. The method `getIterableTypesWithMissingValueTypehint` does not specifically mention `Traversable`, it rests only on `$type->isIterable()->yes()` + `$iterableValue instanceof MixedType && !$iterableValue->isExplicitMixed()`.

Flyingmana commented:
like the tests I added in `tests/PHPStan/Rules/Functions/MissingFunctionReturnTypehintRuleTest.php`? or a more plain phpunit test? (then an example would be helpfull, as I can not find a test which calls processNode() currently)

ondrejmirtes commented:
Yes, it's okay. Thank you!

ondrejmirtes commented:
Please send PRs to the failing extension repositories. Thanks!
